File: backend/app/services/itinerary_service.py
"""Faithful Python port of src/services/itineraryService.ts.

Kept algorithmically identical (same constants, same scoring formulas, same
greedy day-packing loop) so a generated itinerary from this backend matches
what the frontend's own mock generator would have produced for the same
input. Operates on plain dicts (snake_case, matching app/mock_data/*.py and
the schemas in app/schemas/itinerary.py, which accept snake_case via
populate_by_name) rather than ORM/Pydantic objects, so it's easy to unit
test and easy to later swap for a LangGraph pipeline that returns the same
shape.
"""

from datetime import datetime, timezone
from typing import Optional

from app.mock_data.attractions import get_attractions_for
from app.mock_data.destinations import (
    MOCK_DESTINATIONS,
    get_destination_by_id,
    get_destination_by_name,
)
from app.utils.ids import generate_id

DAY_MINUTES_BUDGET = 7 * 60  # ~7 touring hours/day, matches constraints.ts
DAY_START_MINUTES = 9 * 60  # 9:00 AM


def _format_minutes_label(total_minutes: int) -> str:
    hh = (total_minutes % (24 * 60)) // 60
    mm = total_minutes % 60
    period = "PM" if hh >= 12 else "AM"
    hour12 = 12 if hh % 12 == 0 else hh % 12
    return f"{hour12}:{mm:02d} {period}"


def _reason_for(attraction: dict, interests: list[str]) -> str:
    overlap = [c for c in attraction["categories"] if c in interests]
    if overlap:
        label = " and ".join(c[0].upper() + c[1:] for c in overlap)
        return f"Strong match for your interest in {label}."
    return "A well-rated way to experience the destination."


def _relevance_score(attraction: dict, interests: list[str]) -> float:
    if not interests:
        return 50
    overlap = len([c for c in attraction["categories"] if c in interests])
    return (overlap / len(attraction["categories"])) * 100 + overlap * 15


def _generic_attractions_for(destination_name: str, interests: list[str]) -> list[dict]:
    """Fallback used when a freeform / unrecognised destination is given."""
    pool = interests if interests else ["history", "food", "nature", "photography"]
    result = []
    for i, interest in enumerate(pool[:6]):
        result.append(
            {
                "id": f"generic-{destination_name}-{i}",
                "destination_id": "generic",
                "name": f"{interest[0].upper() + interest[1:]} Highlights of {destination_name}",
                "categories": [interest],
                "description": f"A curated local experience centered on {interest} in {destination_name}.",
                "duration_minutes": 90,
                "cost": 250 + i * 50,
                "scene_type": "city",
                "travel_minutes_to_next": 20,
                "distance_km_to_next": 6,
                "best_slot": "morning" if i % 3 == 0 else ("afternoon" if i % 3 == 1 else "evening"),
            }
        )
    return result


def build_itinerary(trip_input: dict, overrides: Optional[dict] = None) -> dict:
    """overrides: {"budget_max": int, "max_hours": float} - used by optimize."""

    overrides = overrides or {}
    destination = get_destination_by_id(trip_input["destination_id"]) or get_destination_by_name(
        trip_input.get("destination_name", "")
    )
    destination_name = (destination or {}).get("name") or trip_input.get("destination_name") or "Your Destination"
    raw_attractions = get_attractions_for(destination["id"]) if destination else []
    pool = raw_attractions if raw_attractions else _generic_attractions_for(destination_name, trip_input["interests"])

    budget_max = overrides.get("budget_max") if overrides.get("budget_max") is not None else trip_input["budget_max"]
    if overrides.get("max_hours") is not None:
        total_minutes_budget = overrides["max_hours"] * 60
    else:
        total_minutes_budget = trip_input["duration_days"] * DAY_MINUTES_BUDGET

    interests = trip_input["interests"]
    scored = sorted(
        [{"attraction": a, "score": _relevance_score(a, interests)} for a in pool],
        key=lambda s: (-s["score"], s["attraction"]["cost"]),
    )

    days = [
        {"day_number": i + 1, "label": f"Day {i + 1}", "summary": "", "stops": []}
        for i in range(trip_input["duration_days"])
    ]

    running_cost = 0
    day_index = 0
    per_day_minutes_cap = total_minutes_budget / trip_input["duration_days"]
    day_cursors = [DAY_START_MINUTES] * len(days)

    for entry in scored:
        attraction = entry["attraction"]
        cursor_idx = day_index % len(days)
        day = days[cursor_idx]
        day_used_minutes = sum(
            s["attraction"]["duration_minutes"] + s["attraction"]["travel_minutes_to_next"] for s in day["stops"]
        )
        stop_minutes = attraction["duration_minutes"] + attraction["travel_minutes_to_next"]

        within_day_time = day_used_minutes + stop_minutes <= per_day_minutes_cap
        within_overall_budget = running_cost + attraction["cost"] <= budget_max * 1.02
        below_min_floor = len(day["stops"]) < 3  # keep every day feeling full even if slightly over budget

        if within_day_time and (within_overall_budget or below_min_floor):
            start_time = _format_minutes_label(day_cursors[cursor_idx])
            day["stops"].append(
                {
                    "attraction": attraction,
                    "time": start_time,
                    "reason": _reason_for(attraction, interests),
                }
            )
            day_cursors[cursor_idx] += stop_minutes
            running_cost += attraction["cost"]
            day_index += 1

    for day in days:
        cats: list[str] = []
        for stop in day["stops"]:
            for c in stop["attraction"]["categories"]:
                if c not in cats:
                    cats.append(c)
        day["summary"] = (
            ", ".join(c[0].upper() + c[1:] for c in cats[:3]) if cats else "Free exploration"
        )

    all_stops = [s for d in days for s in d["stops"]]
    total_cost = sum(s["attraction"]["cost"] for s in all_stops)
    total_touring_minutes = sum(s["attraction"]["duration_minutes"] for s in all_stops)
    total_distance = sum(s["attraction"]["distance_km_to_next"] for s in all_stops)
    place_count = len(all_stops)

    totals = {
        "cost": total_cost,
        "time_hours": round(total_touring_minutes / 60 * 10) / 10,
        "distance_km": round(total_distance),
        "place_count": place_count,
    }

    matched_stops = [s for s in all_stops if any(c in interests for c in s["attraction"]["categories"])]
    if not interests:
        interest_match = 72
    else:
        ratio = len(matched_stops) / max(1, len(all_stops))
        interest_match = min(98, round(ratio * 100 * 0.9 + 15))

    avg_distance = (total_distance / len(all_stops)) if all_stops else 0
    route_efficiency = max(45, min(98, round(100 - avg_distance * 3.2)))

    budget_fit = max(0, min(100, round(100 - ((total_cost - budget_max) / budget_max) * 100))) if budget_max else 0
    available_hours = total_minutes_budget / 60
    time_fit = (
        max(0, min(100, round(100 - ((totals["time_hours"] - available_hours) / available_hours) * 100)))
        if available_hours
        else 0
    )

    scores = {
        "interest_match": interest_match,
        "budget_fit": budget_fit,
        "time_fit": time_fit,
        "route_efficiency": route_efficiency,
    }

    resolved_input = {**trip_input, "destination_name": destination_name}

    return {
        "id": generate_id("itin"),
        "input": resolved_input,
        "days": days,
        "totals": totals,
        "scores": scores,
        "generated_at": datetime.now(timezone.utc).isoformat(),
    }


def generate_itinerary(trip_input: dict) -> dict:
    return build_itinerary(trip_input)


def optimize_itinerary(before: dict, new_budget_max: Optional[int], new_available_hours: Optional[float]) -> dict:
    overrides = {}
    if new_budget_max is not None:
        overrides["budget_max"] = new_budget_max
    if new_available_hours is not None:
        overrides["max_hours"] = new_available_hours
    after = build_itinerary(before["input"], overrides)
    return {
        "before": before,
        "after": after,
        "savings": {
            "cost": max(0, before["totals"]["cost"] - after["totals"]["cost"]),
            "hours": max(0, before["totals"]["time_hours"] - after["totals"]["time_hours"]),
            "distance_km": max(0, before["totals"]["distance_km"] - after["totals"]["distance_km"]),
            "places": max(0, before["totals"]["place_count"] - after["totals"]["place_count"]),
        },
    }


def popular_destination_suggestions() -> list[dict]:
    return [d for d in MOCK_DESTINATIONS if d.get("popular")]

File: frontend/src/services/itineraryService.ts
import type { TripInput, InterestKey } from '@/types/trip'
import type {
  Attraction,
  Itinerary,
  ItineraryDay,
  ItineraryStop,
  ItineraryScores,
  ItineraryTotals,
} from '@/types/itinerary'
import { getAttractionsFor } from '@/data/mockAttractions'
import { getDestinationById, getDestinationByName, mockDestinations } from '@/data/mockDestinations'
import { generateId } from '@/utils/id'
import { simulateLatency } from './api'

const DAY_MINUTES_BUDGET = 7 * 60 // ~7 touring hours/day, matches constraints.ts

const formatMinutesLabel = (totalMinutes: number): string => {
  const hh = Math.floor((totalMinutes % (24 * 60)) / 60)
  const mm = totalMinutes % 60
  const period = hh >= 12 ? 'PM' : 'AM'
  const hour12 = hh % 12 === 0 ? 12 : hh % 12
  return `${hour12}:${mm.toString().padStart(2, '0')} ${period}`
}

const reasonFor = (attraction: Attraction, interests: InterestKey[]): string => {
  const overlap = attraction.categories.filter((c) => interests.includes(c))
  if (overlap.length > 0) {
    const label = overlap.map((c) => c[0].toUpperCase() + c.slice(1)).join(' and ')
    return `Strong match for your interest in ${label}.`
  }
  return `A well-rated way to experience the destination.`
}

const relevanceScore = (attraction: Attraction, interests: InterestKey[]): number => {
  if (interests.length === 0) return 50
  const overlap = attraction.categories.filter((c) => interests.includes(c)).length
  return (overlap / attraction.categories.length) * 100 + overlap * 15
}

const genericAttractionsFor = (destinationName: string, interests: InterestKey[]): Attraction[] => {
  // Fallback used when a freeform / unrecognised destination is typed in.
  const pool: InterestKey[] = interests.length > 0 ? interests : ['history', 'food', 'nature', 'photography']
  return pool.slice(0, 6).map((interest, i) => ({
    id: `generic-${destinationName}-${i}`,
    destinationId: 'generic',
    name: `${interest[0].toUpperCase() + interest.slice(1)} Highlights of ${destinationName}`,
    categories: [interest],
    description: `A curated local experience centered on ${interest} in ${destinationName}.`,
    durationMinutes: 90,
    cost: 250 + i * 50,
    sceneType: 'city',
    travelMinutesToNext: 20,
    distanceKmToNext: 6,
    bestSlot: i % 3 === 0 ? 'morning' : i % 3 === 1 ? 'afternoon' : 'evening',
  }))
}

interface BuildOverrides {
  budgetMax?: number
  maxHours?: number
}

const buildItinerary = (input: TripInput, overrides?: BuildOverrides): Itinerary => {
  const destination = getDestinationById(input.destinationId) ?? getDestinationByName(input.destinationName)
  const destinationName = destination?.name ?? input.destinationName ?? 'Your Destination'
  const rawAttractions = destination ? getAttractionsFor(destination.id) : []
  const pool = rawAttractions.length > 0 ? rawAttractions : genericAttractionsFor(destinationName, input.interests)

  const budgetMax = overrides?.budgetMax ?? input.budgetMax
  const totalMinutesBudget = overrides?.maxHours
    ? overrides.maxHours * 60
    : input.durationDays * DAY_MINUTES_BUDGET

  const scored = [...pool]
    .map((a) => ({ attraction: a, score: relevanceScore(a, input.interests) }))
    .sort((a, b) => b.score - a.score || a.attraction.cost - b.attraction.cost)

  const days: ItineraryDay[] = Array.from({ length: input.durationDays }, (_, i) => ({
    dayNumber: i + 1,
    label: `Day ${i + 1}`,
    summary: '',
    stops: [],
  }))

  let runningCost = 0
  let runningMinutes = 0
  let dayIndex = 0
  const perDayMinutesCap = totalMinutesBudget / input.durationDays
  const DAY_START_MINUTES = 9 * 60 // 9:00 AM
  const dayCursors = new Array(days.length).fill(DAY_START_MINUTES)

  for (const { attraction } of scored) {
    const cursorIdx = dayIndex % days.length
    const day = days[cursorIdx]
    const dayUsedMinutes = day.stops.reduce(
      (sum, s) => sum + s.attraction.durationMinutes + s.attraction.travelMinutesToNext,
      0,
    )
    const stopMinutes = attraction.durationMinutes + attraction.travelMinutesToNext

    const withinDayTime = dayUsedMinutes + stopMinutes <= perDayMinutesCap
    const withinOverallBudget = runningCost + attraction.cost <= budgetMax * 1.02
    const belowMinFloor = day.stops.length < 3 // keep every day feeling full even if slightly over budget

    if (withinDayTime && (withinOverallBudget || belowMinFloor)) {
      const startTime = formatMinutesLabel(dayCursors[cursorIdx])
      day.stops.push({
        attraction,
        time: startTime,
        reason: reasonFor(attraction, input.interests),
      })
      dayCursors[cursorIdx] += stopMinutes
      runningCost += attraction.cost
      runningMinutes += attraction.durationMinutes
      dayIndex++
    }
  }

  days.forEach((day) => {
    const cats = new Set(day.stops.flatMap((s) => s.attraction.categories))
    day.summary = cats.size > 0 ? Array.from(cats).slice(0, 3).map((c) => c[0].toUpperCase() + c.slice(1)).join(', ') : 'Free exploration'
  })

  const totalCost = days.flatMap((d) => d.stops).reduce((sum, s) => sum + s.attraction.cost, 0)
  const totalTouringMinutes = days.flatMap((d) => d.stops).reduce((sum, s) => sum + s.attraction.durationMinutes, 0)
  const totalDistance = days.flatMap((d) => d.stops).reduce((sum, s) => sum + s.attraction.distanceKmToNext, 0)
  const placeCount = days.flatMap((d) => d.stops).length

  const totals: ItineraryTotals = {
    cost: totalCost,
    timeHours: Math.round((totalTouringMinutes / 60) * 10) / 10,
    distanceKm: Math.round(totalDistance),
    placeCount,
  }

  const allStops = days.flatMap((d) => d.stops)
  const matchedStops = allStops.filter((s) => s.attraction.categories.some((c) => input.interests.includes(c)))
  const interestMatch =
    input.interests.length === 0
      ? 72
      : Math.min(98, Math.round((matchedStops.length / Math.max(1, allStops.length)) * 100 * 0.9 + 15))

  const avgDistance = allStops.length > 0 ? totalDistance / allStops.length : 0
  const routeEfficiency = Math.max(45, Math.min(98, Math.round(100 - avgDistance * 3.2)))

  const budgetFit = Math.max(0, Math.min(100, Math.round(100 - ((totalCost - budgetMax) / budgetMax) * 100)))
  const availableHours = totalMinutesBudget / 60
  const timeFit = Math.max(
    0,
    Math.min(100, Math.round(100 - ((totals.timeHours - availableHours) / availableHours) * 100)),
  )

  const scores: ItineraryScores = { interestMatch, budgetFit, timeFit, routeEfficiency }

  return {
    id: generateId('itin'),
    input: { ...input, destinationName },
    days,
    totals,
    scores,
    generatedAt: new Date().toISOString(),
  }
}

export async function generateItinerary(input: TripInput): Promise<Itinerary> {
  await simulateLatency(400)
  return buildItinerary(input)
}

export interface OptimizationResult {
  before: Itinerary
  after: Itinerary
  savings: {
    cost: number
    hours: number
    distanceKm: number
    places: number
  }
}

export async function optimizeItinerary(
  before: Itinerary,
  newBudgetMax: number,
  newAvailableHours: number,
): Promise<OptimizationResult> {
  await simulateLatency(900)
  const after = buildItinerary(before.input, { budgetMax: newBudgetMax, maxHours: newAvailableHours })
  return {
    before,
    after,
    savings: {
      cost: Math.max(0, before.totals.cost - after.totals.cost),
      hours: Math.max(0, before.totals.timeHours - after.totals.timeHours),
      distanceKm: Math.max(0, before.totals.distanceKm - after.totals.distanceKm),
      places: Math.max(0, before.totals.placeCount - after.totals.placeCount),
    },
  }
}

export const popularDestinationSuggestions = () => mockDestinations.filter((d) => d.popular)
